src/main/java/edu/unc/lib/boxc/migration/cdm/services/PermissionsService.java

```diff
@@ -290,27 +291,59 @@ private String getObjectType(String id) {
 
     private List<List<String>> updateCsvRecords(PermissionMappingOptions options) {
         List<CSVRecord> previousRecords = getPermissions();
+        List<Map.Entry<String, String>> workAndFileRecords = new ArrayList<>();
         List<List<String>> updatedRecords = new ArrayList<>();
-        Set<String> cdmIds = new HashSet<>();
+        Set<String> addedAndUpdatedIds = new HashSet<>();
         String everyoneField = getAssignedRoleValue(options.isStaffOnly(), options.getEveryone());
         String authenticatedField = getAssignedRoleValue(options.isStaffOnly(), options.getAuthenticated());
 
-        // update existing entry
+        // addedAndUpdatedIds: list of all ids that need to be added and updated
+        if (options.getCdmId() != null) {
+            addedAndUpdatedIds.add(options.getCdmId());
+        }
+        if (options.isWithWorks() || options.isWithFiles()) {
+            workAndFileRecords = queryForMappedIds(options);
+            Set<String> workFileIds = workAndFileRecords.stream().map(Map.Entry::getKey).collect(Collectors.toSet());
+            addedAndUpdatedIds.addAll(workFileIds);
+        }
+
+        // update existing entries and add unchanged entries to updatedRecords, then remove updated ids from updateIds
         for (CSVRecord record : previousRecords) {
-            cdmIds.add(record.get(0));
-            if (record.get(0).equals(options.getCdmId())) {
+            if (addedAndUpdatedIds.contains(record.get(0))) {
                 updatedRecords.add(Arrays.asList(record.get(0), record.get(1), everyoneField, authenticatedField));
+                addedAndUpdatedIds.remove(record.get(0));
             } else {
                 updatedRecords.add(Arrays.asList(record.get(0), record.get(1), record.get(2), record.get(3)));
+                addedAndUpdatedIds.remove(record.get(0));
             }
         }
 
-        // add new entry
-        if (!cdmIds.contains(options.getCdmId())) {
+        // add one new entry
+        if (addedAndUpdatedIds.contains(options.getCdmId())) {
             String objectType = getObjectType(options.getCdmId());
             updatedRecords.add(Arrays.asList(options.getCdmId(), objectType, everyoneField, authenticatedField));
         }
```

Comment:
I think we can delete this block and only depend on the loop below to add new entries. To do so, we'd update the block at line 301 to also add an entry to `workAndFileRecords` like `workAndFileRecords.put(options.getCdmId(), getObjectType(options.getCdmId()))`. You might need to change line 305 to add the results from `queryForMappedIds` to `workAndFileRecords` instead of setting the variable to the result, otherwise the cdmId entry would be lost.